x-pack/plugin/otel: introduce x-pack-otel plugin

gregkalapos opened the pull request:
Solves https://github.com/elastic/elasticsearch/issues/104731

This PR adds mappings for OpenTelemetry data via a new plugin `otel-data`

The PR is mainly based on:
- the existing apm-data plugin from https://github.com/elastic/elasticsearch/pull/97546
- the https://github.com/elastic/elasticsearch/pull/104455 draft PR by @felixbarny - most of the mappings are from there

What happens in the PR:
- Created `YamlTemplateRegistry` as an abstract base class that loads mappings from YAML. The code of this class comes from `APMIndexTemplateRegistry` which now derives from `YamlTemplateRegistry`. The shared code is now moved to `core`.
- Added `otel-data` plugin with `OtelIndexTemplateRegistry` which also derives from `YamlTemplateRegistry`.
- Moved the mappings into YAML files from https://github.com/elastic/elasticsearch/pull/104455.

TODOs:
- [x] Discuss if we want to share some mappings between APM and OTel (candidates: all traces@ files from https://github.com/elastic/elasticsearch/pull/104455), those are currently not yet in the PR -> ~Discussed: not at this point~ Update: we did that - now some some basic things are shared across otel and apm.
- [x] Discuss data stream lifecycle for OTel mappings (right now, nothing is set in the PR) -> Discussed: Since these mappings are shared across multiple OTel use-cases (e.g. system metrics and APM, for now we don't set a default ILM/DSL, since it's hard to come up with a default. Using what APM does today may not be good for integrations - so for now no DSL is set in these templates).
- [x] Adding more tests to `otel-data`

gregkalapos committed:
Add YamlTemplateRegistry and OtelIndexTemplateRegistry with resource YAML files

gregkalapos committed:
Fix traces-otel template

gregkalapos committed:
Adding first yml tests

gregkalapos committed:
Base APMIndexTemplateRegistry on YamlTemplateRegistry

gregkalapos committed:
Update OTelPlugin.java

gregkalapos committed:
Update APMIndexTemplateRegistry.java

gregkalapos committed:
Update YamlIngestPipelineConfig.java

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/ecs-tsdb@mappings.yaml:
```diff
@@ -0,0 +1,20 @@
+template:
+  mappings:
+    dynamic_templates:
+      - ecs_ip:
+          mapping:
+            type: ip
+          path_match:
+            - "ip"
+            - "*.ip"
+            - "*_ip"
+          match_mapping_type: string
```
Because of https://github.com/elastic/elasticsearch/issues/110387, we can't map all `*.ip` fields to the  `ip` field type.

```suggestion
```

gregkalapos commented on x-pack/plugin/otel-data/src/main/resources/component-templates/traces-otel@mappings.yaml:
```diff
@@ -0,0 +1,71 @@
+        properties:
+          message:
+            type: keyword
+          status_code:
```
In the table in our doc this is called `status.code` and not `status.status_code`: https://github.com/elastic/opentelemetry-dev/blob/main/docs/design-decisions/ingest/mapping.md#traces

`status.status_code` comes from the original draft PR.

We should get this in sync.

My proposal: I think `status.code` is better, I see no reason for the redundancy, so similar to `message` I'd just call it `status_code`. 

Is that ok, or was there any other reason for this?

felixbarny replied:
Good catch. +1 to changing this to `status.code`.

gregkalapos committed:
Adding traces tests

gregkalapos committed:
Update x-pack/plugin/otel-data/src/main/resources/component-templates/ecs-tsdb@mappings.yaml

Co-authored-by: Felix Barnsteiner <felixbarny@users.noreply.github.com>

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/otel@mappings.yaml:
```diff
@@ -0,0 +1,67 @@
+            ignore_above: 1024
+# TODO:
+#Causes error during YAML test
+# [2024-07-19T04:36:51,833][ERROR][o.e.x.c.t.IndexTemplateRegistry] [test-cluster-0] error adding index template [otel@mappings] for [otel] org.elasticsearch.index.mapper.MapperParsingException: Failed to parse mapping: No handler for type [version] declared on field [version]
```
You'll need to add a test dependency on the `mapper-version` module. See also

https://github.com/elastic/elasticsearch/blob/ba3501ae29b010643577379446bc49fdf4775320/x-pack/plugin/profiling/build.gradle#L23

felixbarny commented on x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/OTelPlugin.java:
```diff
@@ -0,0 +1,79 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.oteldata;
```
I'm wondering whether we should call the plugin just `otel`. That may be more future-proof in case we want to also add something like an OTLP endpoint to ES at some point. Having said that, `otel-data` may still be fitting and if not we could also just rename the module.

axw replied:
FWIW the "apm-data" plugin was originally just called "x-pack-apm", and was renamed because it introduced confusion with modules/apm. I could see the same thing happening with otel, though maybe if you get in first it doesn't matter ;)

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/metrics-otel@mappings.yaml:
```diff
@@ -0,0 +1,55 @@
+        type: keyword
+        time_series_dimension: true
+        ignore_above: 1024
+    dynamic_templates:
```
Let's move these dynamic templates to `metrics@mappings`, so that we can have a simpler logic for shippers so that they can use the same dynamic templates regardless of which metric data stream is targeted.

felixbarny replied:
Based on a quick check, it seems `time_series_metric` gets ignored when targeting a non tsdb data stream.

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/otel@mappings.yaml:
```diff
@@ -0,0 +1,67 @@
+    dynamic: strict
+    properties:
+      "@timestamp":
+        type: date_nanos
```
Because ES|QL doesn't support `date_nanos` at the moment (see https://github.com/elastic/elasticsearch/issues/109352), I think we need to revert to using `date` for now, until ES|QL reaches an acceptable level of support for `date_nanos`.

@mlunadia @strawgate @tylerperk, we could use your help to define what acceptable means and what level of parity we need in order for us to be ready to use the `date_nanos` field type. Do we need the experience with `date_nanos` to not be worse in any aspects compared to `date`? Are we willing to make compromises, for example users having to make manual casts in ES|QL? What are the essentials and what are the nice to haves?

strawgate replied:
Would it make sense to default to `date` here but override it to `date_nanos` for specific datasets? I think most integrations do not use `date_nanos` today and it carries with it perf/storage costs?

felixbarny replied:
They don't but we're also discussing changing this as a default for integrations: https://github.com/elastic/logs-dev/issues/25

There's minimal overhead if `date_nanos` is used as a field type but the timestamps are only in ms precision. As OTel timestamps are defined in [nanosecond in OTLP](https://github.com/open-telemetry/opentelemetry-proto/blob/484241a016d7b81f20b6e19d00ffbc4a3b864a22/opentelemetry/proto/logs/v1/logs.proto#L140), I think we should also map them this way, without losing precision, by choosing the `date_nanos` field type by default.

Users can still override this and chose `date` instead.

However, in the short term, because of the limitations in ES|QL, I don't think we can use the `date_nanos` field type, yet.

gregkalapos committed:
Add mapper-version

gregkalapos committed:
Fix code-style

gregkalapos committed:
Rename `status.status_code` to `status.code`

gregkalapos committed:
Update otel@mappings.yaml

Revert back to date due to missing support in ES|QL for date_nanos

gregkalapos committed:
Move dynamic_templates to metrics@mappings in core

gregkalapos commented on x-pack/plugin/otel-data/src/main/resources/index-templates/logs-otel@template.yaml:
```diff
@@ -0,0 +1,21 @@
+  description: default logs template installed by x-pack
+  managed: true
+composed_of:
+  - logs@mappings
```
This would allow us to share very basic mappings and not duplicate those. 

In this specific one, the only thing that we pull in is to set [`data_stream.type` to ` "value": "logs"`](https://github.com/elastic/elasticsearch/blob/86c41c6670d72986ac07116bf911e8c274b10a6b/x-pack/plugin/core/template-resources/src/main/resources/logs%40mappings.json#L11) - the remaining things from `logs@mappings` are already covered in `otel@mappings`. I'd say setting `data_stream.type` to ` "value": "logs"` is something which makes sense to share across log related templates.

I don't have much history on how often these mappings from core are changed - or if there would be any risk here, but I think this would be a good way to share these very basic mappings and also use it for OTel.

We discussed that potentially we could also share a few things for traces between otel and `apm-data` - that'd be also done via a mapping in `core`. But that's outside this PR.

gregkalapos replied:
That's an outcome from this discussion: https://github.com/elastic/elasticsearch/pull/111091#discussion_r1710769086.

gregkalapos committed:
Run gradlew :x-pack:plugin:core:spotlessApply

carsonip commented on x-pack/plugin/otel-data/src/main/resources/component-templates/metrics-otel@mappings.yaml:
```diff
@@ -0,0 +1,20 @@
+      data_stream.type:
+        type: constant_keyword
+        value: metrics
+      start_time:
```
```suggestion
      start_timestamp:
```

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/traces-otel@mappings.yaml:
```diff
@@ -0,0 +1,71 @@
+template:
+  mappings:
+    date_detection: false
+    dynamic: strict
```
Redundant as it's set in otel@mappings.

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/logs-otel@mappings.yaml:
```diff
@@ -0,0 +1,41 @@
+template:
+  mappings:
+    date_detection: false
+    dynamic: strict
```
Redundant as it's set in otel@mappings.

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/otel@mappings.yaml:
```diff
@@ -0,0 +1,65 @@
+template:
+  mappings:
+    date_detection: false
+    dynamic: strict
```
Let's change this to `dynamic: false` to avoid forward compatibility issues.

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/ecs-tsdb@mappings.yaml:
```diff
@@ -0,0 +1,12 @@
+  managed: true
+template:
+  mappings:
+    dynamic_templates:
```
Let's add a dynamic template that maps boolean fields to keywords. This is to work around https://github.com/elastic/elasticsearch/issues/111338. I think that's preferable over converting booleans to strings in the ES exporter. That's because removing the conversion in the ES exporter would require it to check whether or not the connected ES cluster supports boolean dimensions. When we change it in the dynamic template, the ES exporter doesn't need to know about that and it can always assume that booleans are supported.

cc @carsonip

carsonip replied:
+1. If that's ready, I'll remove the workaround in my draft PR.

QQ: what's the implication on backcompat? If a doc is ingested to 8.16 with the workaround in dynamic template, and 8.17 removed that workaround, will it cause any issues at query time?

felixbarny replied:
Not quite sure. Maybe you need to adjust term queries to drop quotes like `"value": "true"` -> `"value": true`. But I'm not too concerned about that as long as this isn't GA. Maybe ES can already coerce this at query time which would make it a complete non-issue then.

gregkalapos replied:
Pushed https://github.com/elastic/elasticsearch/pull/111091/commits/a9987d38ce15b2fee5674a21974fbdc3da1b5d6b

- Added a test that has `transaction.root: false` that triggers the error `Error extracting routing: Routing values must be strings but found [VALUE_BOOLEAN]`
- Added a dynamic mapping to map boolean to keyword

However that mapping does not work - discussed with @felixbarny, problem is likely that routing happens before dynamic mapping would kick in. Therefore, I commented out that mapping and it will be likely removed from the PR.

For now, the test is also skipped, since the dynamic mapping is not in place.

After slacking with @felixbarny, there is https://github.com/elastic/elasticsearch/pull/111445 open. If that gets in, there won't be a need for `booleans_to_keywords` dynamic mapping.

Alternatively, if https://github.com/elastic/elasticsearch/pull/111445 won't get in, we'll need to look for another approach.

carsonip replied:
As https://github.com/elastic/elasticsearch/pull/111445 is merged, can you rebase (merge main) this PR, so that it is available in a new custom docker build?

gregkalapos replied:
👍 

We still need `booleans_to_keywords`, I enabled that and also removed skipping the corresponding test which after rebasing on `main` runs through successfully. So it seems to be working and ready to test.

gregkalapos committed:
Update x-pack/plugin/otel-data/src/main/resources/component-templates/metrics-otel@mappings.yaml

Co-authored-by: Carson Ip <carsonip@users.noreply.github.com>

gregkalapos committed:
Update 20_metic_tests.yml

Workaround for TSDB timestamp issue: we push a custom template with higher priority and set  time_series.start_time.

gregkalapos committed:
Update CODEOWNERS

Adding obs-ds-intake-services as owner of the new otel-data plugin.

Since we had some changes, also updating the owner of apm-data to the same team.

gregkalapos committed:
Change dynamic: strict to false

gregkalapos commented on .github/CODEOWNERS:
```diff
@@ -27,8 +27,12 @@ libs/logstash-bridge @elastic/logstash
 # APM Data index templates, etc.
-x-pack/plugin/apm-data/src/main/resources @elastic/apm-server
-x-pack/plugin/apm-data/src/yamlRestTest/resources @elastic/apm-server
+x-pack/plugin/apm-data/src/main/resources @elastic/obs-ds-intake-services
+x-pack/plugin/apm-data/src/yamlRestTest/resources @elastic/obs-ds-intake-services
+
+# OTel
+x-pack/plugin/otel-data/src/main/resources @elastic/obs-ds-intake-services
+x-pack/plugin/otel-data/src/yamlRestTest/resources @elastic/obs-ds-intake-services
```
@felixbarny talked to the intake service team - I was told both for the new otel plugin we should add this team, plus we should update `apm-data` to be owned by this team as well. It seems we need to sort out GitHub permissions

gregkalapos commented on x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/20_metic_tests.yml:
```diff
@@ -0,0 +1,48 @@
+
+---
+"Test push service overview metric":
+  - do:
+      indices.put_index_template:
+        name: test@template
+        body:
+          priority: 500
+          index_patterns: [ "metrics-*.otel-*" ]
+          composed_of:
+            - metrics@tsdb-settings
+            - otel@mappings
+            - metrics@mappings
+            - metrics-otel@mappings
+            - semconv-resource-to-ecs@mappings
+            - ecs-tsdb@mappings
+          template:
+            settings:
+              index:
+                routing_path: [unit, attributes.*, resource.attributes.*]
+                mode: time_series
+                time_series:
+                  start_time: 2024-07-01T13:03:08.138Z
```
The challenge here is that we need to set `time_series.start_time`, or use dynamic `@timestamp` to avoid rejecting too old `@timestamp` by TSDB.

I tried:
- Updating the existing `metrics-otel@template`, which would be the ideal solution. If I just use `indices.put_index_template` with the existing `metrics-otel@template` name, it gets overwritten, so does not seem to work.
- Tried to push a `metrics-otel@custom` mapping - I was not able to figure out how to do that with this `yml` syntax. Haven't found any sample in the repo and it got very painful to test it over and over. This would be still fine in my opinion, so if someone knows how to do so, it'd be great.
- Alternatively we could look into how to use dynamic `@timestamps` would be easy in Java - or in any other language - but so far I have not found a way to do that in `yml`.

With that, the remaining idea was to re-create the whole mapping with setting `time_series.start_time`. But overall I think this is not very good.

felixbarny replied:
> Tried to push a metrics-otel@custom mapping - I was not able to figure out how to do that with this yml syntax. 
When you search for `cluster.put_component_template`, you can find a bunch of examples.

It should be something similar to 

```yml
  - do:
      cluster.put_component_template:
        name: metrics-otel@custom
        body:
          template:
            settings:
              index:
                time_series:
                  start_time: 2024-07-01T13:03:08.138Z
```

If setting the start_time doesn't work here, I have another idea which involves setting the `@timestamp` in an ingest pipeline.

```yml
  - do:
      ingest.put_pipeline:
        id: "default_timestamp"
        body:  >
          {
            "description": "_description",
            "processors": [
              {
                "set": {
                  "description": "If '@timestamp' is missing, set it with the ingest timestamp",
                  "field": "@timestamp",
                  "override": false,
                  "copy_from": "_ingest.timestamp"
                }
              }
            ]
          }
  - match: { acknowledged: true }

  - do:
      cluster.put_component_template:
        name: metrics-otel@custom
        body:
          template:
            settings:
              index:
                default_pipeline: default_timestamp
```

gregkalapos replied:
Nice, thanks! Pushing a custom `metrics-otel@custom` worked - I think I was searching for `indices.put_component_template` and not `cluster.put_component_template`. Anyways, this works.

gregkalapos committed:
Skip "Reject invalid top level field" test

gregkalapos committed:
Update 20_metic_tests.yml

gregkalapos committed:
Add boolean as dimension test (skipping it for now)

felixbarny commented:
@elasticmachine update branch

felixbarny commented:
@elasticsearchmachine update branch

felixbarny commented:
@elasticmachine update branch

elasticmachine commented:
expected head sha didn’t match current head ref.

gregkalapos committed:
Add booleans_to_keywords and enable corresponding test

carsonip commented on x-pack/plugin/otel-data/src/main/resources/component-templates/otel@mappings.yaml:
```diff
@@ -0,0 +1,65 @@
+        time_series_dimension: true
+      dropped_attributes_count:
+        type: long
+      processor.event:
```
Remove `processor.event` to allow attributes.processor.event

gregkalapos replied:
Done in https://github.com/elastic/elasticsearch/pull/111091/commits/5b4b65f71825eb40ab84fd3caead07506c74d409

gregkalapos committed:
Remove processor.event top level mapping

Reason: for metrics and logs we can rely on the name of the datastream. For spans vs. transactions there are other fields we can use.

gregkalapos committed:
Remove booleans_to_keywords

Because booleans are supported now as dimension on TSDB

gregkalapos committed:
Add alias service.language.name -> telemetry.sdk.language

gregkalapos committed:
cleanup

breskeby commented on x-pack/plugin/otel-data/README.md:
````diff
@@ -0,0 +1,44 @@
+These can be run with:
+
+```
+./gradlew x-pack:plugin:apm-data:yamlRestTest
````
```suggestion
./gradlew :x-pack:plugin:apm-data:yamlRestTest
```

gregkalapos replied:
This part got removed -  https://github.com/elastic/elasticsearch/pull/111091#discussion_r1704441078

gregkalapos committed:
Update README.md

breskeby commented on x-pack/plugin/otel-data/README.md:
````diff
@@ -0,0 +1,44 @@
+These can be run with:
+
+```
+./gradlew x-pack:plugin:apm-data:test
````
```suggestion
./gradlew :x-pack:plugin:apm-data:test
```

gregkalapos replied:
Thanks, fixed this - also this should have been `:x-pack:plugin:otel-data:test`, fixed that as well.

breskeby commented on x-pack/plugin/otel-data/README.md:
````diff
@@ -0,0 +1,44 @@
+These can be run with:
+
+```
+./gradlew x-pack:plugin:apm-data:test
````
why are we referencing unit tests in this plugins readme but tests live in apm-data?

gregkalapos replied:
We don't. Totally my fault, this was some leftover. 

I noticed just right after I flipped this PR to ready-to-review, then I pushed https://github.com/elastic/elasticsearch/pull/111091/commits/1c6192990883db59473710541e08489c679f33b2 - probably the same time as you looked at it. Sorry about this.

breskeby reviewed (commented):
build related changes lgtm. two remarks / questions though

gregkalapos committed:
Update README.md

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/traces@mappings.yaml:
`traces@mappings` already exists in `apm-data/src/main/resources/component-templates/traces@mappings.yaml` so there will be a race about which index template is installed.
Let's instead move `traces@mappings` and `traces@settings` to `x-pack/plugin/core/template-resources/src/main/resources` (and convert it to JSON along the way). Make sure to adjust `StackTemplateRegistry` to pick up the file.

gregkalapos replied:
Makes sense - refactored this in https://github.com/elastic/elasticsearch/pull/111091/commits/f42756445d4e409b12d23a26ed85fa3b7a6508a0.

felixbarny replied:
What's missing is to remove the duplicate `traces@mapings` from the apm-data plugin.

gregkalapos replied:
See https://github.com/elastic/elasticsearch/pull/111091#discussion_r1705898022

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/traces@settings.yaml:
```diff
@@ -0,0 +1,12 @@
+  managed: true
+template:
+  settings:
+#    index:
+#      lifecycle:
+#        name: traces
```
```suggestion
```

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/index-templates/traces-otel@template.yaml:
```diff
@@ -0,0 +1,20 @@
+  description: default OpenTelemetry traces template installed by x-pack
+  managed: true
+composed_of:
+  - otel@mappings
```
Add `traces@settings` and `traces@mappings`?

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/resources.yaml:
```diff
@@ -0,0 +1,16 @@
+# pipeline is changed, in order for it to be updated on Elasticsearch upgrade.
+version: 1
+
+component-templates:
```
This is missing `traces@mappings` and `traces@settings` but these should be moved to the core anyway.

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/semconv-resource-to-ecs@mappings.yaml:
```diff
@@ -0,0 +1,133 @@
+        path: resource.attributes.k8s.pod.uid
+# Below is a non-ECS field, ideally we should have this in a separate file
+# I'd imagine having 3 files:
+# 1) for the fixed mapping for known resource.attributes.* fields that we want to alias
+# 2) this file that aliases ECS -> SemConv
+# 3) a new file that aliases non-ECS fields the UI uses -> SemConv
```
The alias fields need to be defined in the same component template as the fields that they target. That's because each component template is validated individually (in addition to the result of the composition). I don't think that we need to be super strict about splitting ECS and not technically ECS fields. Technically speaking, the `kubernetes.*` fields are also not part of ECS, yet I don't think it makes much sense to have a separate file for it.

Bottom line is that I think it's fine the way it is.

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/index-templates/metrics-otel@template.yaml:
```diff
@@ -0,0 +1,25 @@
+  description: default OpenTelemetry metrics template installed by x-pack
+  managed: true
+composed_of:
+  - metrics@tsdb-settings
+  - otel@mappings
```
Let's be consistent with the ordering of whether settings or mappings are listed first. Also have a look at the index templates in core to ensure consistency with them.

felixbarny replied:
IIRC, in the apm-data plugin, there are unit tests that verify the consistency of the different index templates. It probably makes sense to adapt them for this plugin, too. Feel free to do this as a follow-up.

gregkalapos replied:
Seems it's always first mappings and then settings - so I aligned it. I'll defer the unit tests to a follow up PR - or if we need to wait longer for a review here I may push it here - it's on my todo list.

felixbarny commented on x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/20_metic_tests.yml:
```diff
@@ -0,0 +1,88 @@
+        refresh: true
+        body:
+          - create: {}
+          - '{"@timestamp":"2024-07-18T14:48:33.467654000Z","data_stream":{"dataset":"generic.otel","namespace":"default","type":"metrics"},"processor":{"event":"metric"},"resource":{"attributes":{"service.name":"OtelSample","telemetry.sdk.language":"dotnet","telemetry.sdk.name":"opentelemetry"},"dropped_attributes_count":0}}'
```
This still mentions processor.event at the top level.

gregkalapos replied:
Ah, thanks! moved under `attributes`.

felixbarny commented on x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/20_metic_tests.yml:
```diff
@@ -0,0 +1,88 @@
+        body:
+          fields: ["service.name", "telemetry.sdk.language", "telemetry.sdk.name" ]
+  - length: { hits.hits: 1 }
+  - match: { hits.hits.0.fields: {"service.name": ["OtelSample"], "telemetry.sdk.language": ["dotnet"], "telemetry.sdk.name": ["opentelemetry"] }}
```
I'd prefer this in terms of readability and consistency with other tests in the codebase
```suggestion
  - match: { hits.hits.0.fields.service\.name:  ["OtelSample"] }  
  - match: { hits.hits.0.fields.telemetry\.sdk\.language:  ["dotnet"] }
  ...
```

Also applies to other places in this and other test files where assertions on `fields` are made.

gregkalapos replied:
Pushed it in https://github.com/elastic/elasticsearch/pull/111091/commits/d5de2a1625866f227c84efd3c44b198dc2659a65.

felixbarny commented on x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/20_traces_tests.yml:
```diff
@@ -0,0 +1,71 @@
+  - length: { hits.hits: 1 }
+
+---
+"Reject invalid top level field":
```
Can be removed. Maybe change the test to verify that there are no `fields` in the search response for unknown top-level fields.

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/logs-otel@mappings.yaml:
We should follow up soon to enable logsdb for OTel logs (`settings.index.mode: logsdb`). For now, we can start by enabling [synthetic source](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-source-field.html).

strawgate replied:
Should we just do that right away in this PR?

felixbarny replied:
I'd like to optimize for getting the initial version in soon so that we can unblock further testing and find out if there are any other UI-related issues that we need to prioritize getting fixed.

Greg will create an issue with a list of follow-up tasks, where enabling LogsDB is one of the tasks. We'll need to define which fields we want to sort by default, which is an important factor for how much storage can be saved. We'll then also need to add these fields to the mapping statically as fields used in index sorting need to be defined in the mapping upfront. In its easiest form, we can just add `resource.attributes.host.name` to the mapping and define that in `index.sort.field`. Ideally, we'll want to sort by all resource attributes (see https://github.com/elastic/elasticsearch/issues/110792). I guess we can start simple with the host name, then discuss adding a few more common resource attributes (some candidates could be `host.id`, `service.name`, `service.instance.id`, `k8s.pod.uid`), and finally just incorporate all resource attributes once ES supports that.

gregkalapos replied:
Pushed a patch from @felixbarny that he sent over slack: https://github.com/elastic/elasticsearch/pull/111091/commits/ef3468f5f6d61b0c4c32f7f1a34698762ad365e1

I opened https://github.com/elastic/elasticsearch/issues/111877 - I don't tick this item with the commit above, so we keep this open to think more about other `index.sort.field` fields.

strawgate replied:
Amazing!

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/otel@mappings.yaml:
```diff
@@ -0,0 +1,63 @@
+template:
+  mappings:
+    date_detection: false
+    dynamic: false
```
Let's enable synthetic _source for all otel signals from the beginning. Based on the recent enhancements to synthetic _source, there shouldn't be any remaining blockers.

```suggestion
    dynamic: false
    _source:
      mode: synthetic
```

felixbarny reviewed (commented):
Let's enable synthetic _source :)

gregkalapos committed:
Update docs/changelog/111091.yaml

carsonip commented on x-pack/plugin/otel-data/src/main/resources/component-templates/traces-otel@mappings.yaml:
```diff
@@ -0,0 +1,70 @@
+  mappings:
+    date_detection: false
+    properties:
+      data_stream.namespace:
```
```suggestion
      data_stream.type:
```

carsonip replied:
Also why is this specified here, if it is already specified in traces@mappings.yaml?

gregkalapos replied:
Thanks for catching this. 

> Also why is this specified here, if it is already specified in [traces@mappings.yaml](mailto:traces@mappings.yaml)?

I suspect that is because here we set `value: traces`, which was not (or should have not been) present on `traces@mappings`. For now I just removed this part. Let's discuss if we want to have it here or in `traces@mappings`.

gregkalapos committed:
Move traces@settings and traces@mappings to core

gregkalapos committed:
Update traces-otel@mappings.yaml

gregkalapos commented on x-pack/plugin/core/template-resources/src/main/resources/traces@mappings.json:
```diff
@@ -0,0 +1,28 @@
+        },
+        "data_stream.type": {
+          "type": "constant_keyword",
+          "value": "traces"
```
Are we ok with this here or should we move that to `traces-otel@mappings`? Related: https://github.com/elastic/elasticsearch/pull/111091#discussion_r1705736505

APM uses the same value, doesn't it?

felixbarny replied:
Yes, it's fine to have it here.

felixbarny replied:
We should also delete the `traces@mappings` in APM data and rely on this one instead.

gregkalapos replied:
> We should also delete the `traces@mappings` in APM data and rely on this one instead.

Yeah, that was my plan. 

I did that here: https://github.com/elastic/elasticsearch/pull/111091/commits/f42756445d4e409b12d23a26ed85fa3b7a6508a0#diff-6a6ef4571ed7574fdc81c6f5bdaadce140dc08c70ff2c947394f71b604d2169b

Did I overlook something?

felixbarny replied:
Sorry, I overlooked that.

lahsivjar commented on x-pack/plugin/otel-data/src/main/resources/component-templates/semconv-resource-to-ecs@mappings.yaml:
```diff
@@ -0,0 +1,133 @@
+              process.executable.path:
+                type: keyword
+                ignore_above: 1024
+              process.runtime.path:
```
```suggestion
              process.runtime.name:
```
I think this should be `process.runtime.name`

carsonip commented on x-pack/plugin/otel-data/src/main/resources/component-templates/traces-otel@mappings.yaml:
```diff
@@ -0,0 +1,67 @@
+        type: alias
+        path: parent_span_id
+      trace_flags:
+        type: byte
```
type `byte` is probably not enough. See https://github.com/open-telemetry/opentelemetry-collector/blob/main/pdata/internal/data/protogen/trace/v1/trace.pb.go#L400-L421

gregkalapos replied:
What type do we want to use here?

The collector uses unsigned int, but we don't have unsigned int in Elasticsearch.

I see 2 options:
- integer
  - Pro: less space - we only use 32bit
  - Con: Not future proof. The comment on span indicates that the 10 least significant bits are used, so that means, we can store those values in a signed integer. But then it also says: `Readers MUST NOT assume that bits 10-31 (22 most significant bits) will be zero.` - now that means this may already overflow if we store it in a signed int. Furthermore, Logs has the same field (I'll update that field as well), and [that comment specifically says](https://github.com/open-telemetry/opentelemetry-collector/blob/main/pdata/internal/data/protogen/logs/v1/logs.pb.go#L406-L411): `24 most significant bits are reserved`.

- long
  - Con: it's a 64bit field, so we waste 32 bit
  - Pro: if most significant bits are getting used in the collector uint, we can still store those values. So this addresses the `cons` with integer.

I'd note that an alternative approach could be to parse these flags in the collector (probably in the es exporter) and store those in separate fields. Currently there are 2 things there:
- The 8bit trace-flag from the W3C traceparent header - [that has currently 2 bits defined](https://www.w3.org/TR/trace-context-2/#sampled-flag) - sampled and  random trace id.
- In case of a span, whether the parent is remote, which we just removed: https://github.com/elastic/opentelemetry-dev/pull/364.

I'm not sure parsing is the way to go, the reason I bring it up is because these bitflag in its current form may not be the most useful. I don't really see what we'd do with it as is.

Overall, I'd suggest keeping the field and using type `long`.

What do others think? @felixbarny.

gregkalapos replied:
I did a quick test to see what happens if those most significant bits are set and we use integer.

Elasticsearch won't let it overflow and set it to a negative number - so even if in theory this would work, Elasticsearch prevents us doing it - which kind of makes sense. So this really makes me think we should go with long here.

```
{
  ".ds-foobar-2024.08.08-000001": {
    "mappings": {
      "_data_stream_timestamp": {
        "enabled": true
      },
      "properties": {
        "@timestamp": {
          "type": "date"
        },
        "intfield": {
          "type": "integer"
        }
      }
    }
  }
}
```

```
POST foobar/_doc
{
  "@timestamp":  "2024-08-05T10:43:36.250Z",
  "intfield": 4294967295
}
```


Response:

```
{
  "error": {
    "root_cause": [
      {
        "type": "document_parsing_exception",
        "reason": "[1:59] failed to parse field [intfield] of type [integer] in document with id 'ZDKrMpEBZ0KOL9if4h2U'. Preview of field's value: '4294967295'"
      }
    ],
    "type": "document_parsing_exception",
    "reason": "[1:59] failed to parse field [intfield] of type [integer] in document with id 'ZDKrMpEBZ0KOL9if4h2U'. Preview of field's value: '4294967295'",
    "caused_by": {
      "type": "x_content_parse_exception",
      "reason": """[1:69] Numeric value (4294967295) out of range of int (-2147483648 - 2147483647)
 at [Source: (byte[])"{  "@timestamp": "2024-08-05T10:43:36.250Z",  "intfield": 4294967295}
"; line: 1, column: 69]""",
      "caused_by": {
        "type": "input_coercion_exception",
        "reason": """Numeric value (4294967295) out of range of int (-2147483648 - 2147483647)
 at [Source: (byte[])"{  "@timestamp": "2024-08-05T10:43:36.250Z",  "intfield": 4294967295}
"; line: 1, column: 69]"""
      }
    }
  },
  "status": 400
}
```

carsonip replied:
`long` might be the only option if we want to be compliant and avoid fixing this down the line when otel decides to start using the significant bits.

felixbarny replied:
Trace flags is an 8 bit field: https://www.w3.org/TR/trace-context/#trace-flags
Why do we need another data type than byte?

gregkalapos replied:
> Trace flags is an 8 bit field: https://www.w3.org/TR/trace-context/#trace-flags

That is just one part of what's stored in the [`flags` field linked above](https://github.com/open-telemetry/opentelemetry-collector/blob/main/pdata/internal/data/protogen/trace/v1/trace.pb.go#L400-L421). The first 8 bits of the `flags` on the span is the 8bit trace flags from W3C, then the next 2 bits are the `is remote parent` thing (with it's 3 values), and the remaining 22 bits are not used, but as the comment indicates may not be zero.

So this is a flag, which stores another flag in its first part :)

gregkalapos replied:
> `long` might be the only option if we want to be compliant and avoid fixing this down the line when otel decides to start using the significant bits.

Not only when they start using the significant bits. Sure, that's a clear problem then. But even if those are not zeroed, as the comment indicates (even if there is no definition for those yet), then Elasticsearch will already reject those numbers.

felixbarny replied:
Oh, I see. Thanks for the clarification. While ES does have an [`unsigned_long`](https://www.elastic.co/guide/en/elasticsearch/reference/current/number.html) data type, there's no `unsigned_integer`. Therefore, `long` seems to be the best fit.

gregkalapos replied:
Thanks for the input - seems we have an agreement.

Changed it to `long`: https://github.com/elastic/elasticsearch/pull/111091/commits/a6baf4644924ea4929ae2ab58ebac0b78f936e62

And a PR to document this in our spec: https://github.com/elastic/opentelemetry-dev/pull/368

gregkalapos replied:
Update: we decided to drop this field for now, so it's removed from the mappings. See: https://github.com/elastic/opentelemetry-dev/pull/371

carsonip commented on x-pack/plugin/otel-data/src/main/resources/component-templates/traces-otel@mappings.yaml:
```diff
@@ -0,0 +1,67 @@
+          dropped_attributes_count:
+            type: long
+          trace_flags:
+            type: byte
```
ditto

carsonip commented on x-pack/plugin/otel-data/src/main/resources/component-templates/traces-otel@mappings.yaml:
```diff
@@ -0,0 +1,67 @@
+        path: parent_span_id
+      trace_flags:
+        type: byte
+      parent_is_remote:
+        type: boolean
```
```suggestion
```
Removing parent_is_remote as discussed on [slack](https://elastic.slack.com/archives/C0680QCJKMM/p1723039068008969?thread_ts=1722953805.258019&cid=C0680QCJKMM)

carsonip replied:
ref: https://github.com/elastic/opentelemetry-dev/pull/364

gregkalapos committed:
Review feedback

gregkalapos committed:
Adapt `match` style in tests

carsonip commented on x-pack/plugin/otel-data/src/main/resources/component-templates/traces-otel@mappings.yaml:
```diff
@@ -0,0 +1,67 @@
+        path: parent_span_id
+      trace_flags:
+        type: byte
+      parent_is_remote:
+        type: boolean
```
```suggestion
```
Remove parent_is_remote according to change in spec

gregkalapos committed:
Update docs/changelog/111091.yaml

elasticsearchmachine commented:
Hi @GregKalapos, I've updated the changelog YAML for you.

gregkalapos commented:
> Hi @gregkalapos, I've updated the changelog YAML for you.

Why?! I tried to be a proper contributor and added the related issue - now the bot removes it. :(

gregkalapos committed:
Apply suggestions from code review

Co-authored-by: Vishal Raj <vishal.raj@elastic.co>

gregkalapos commented on x-pack/plugin/otel-data/src/main/resources/component-templates/semconv-resource-to-ecs@mappings.yaml:
```diff
@@ -0,0 +1,133 @@
+        path: resource.attributes.process.executable.path
+      service.runtime.name:
+        type: alias
+        path: resource.attributes.process.runtime.path
```
Good catch, yes. There is `process.executable.path`, which we already handle - but there is no `process.runtime.path`. And the alias should definitely point to `process.runtime.name` - `process.executable.path` would be different.

gregkalapos committed:
Update x-pack/plugin/otel-data/src/main/resources/component-templates/traces-otel@mappings.yaml

Co-authored-by: Carson Ip <carsonip@users.noreply.github.com>

carsonip commented on x-pack/plugin/otel-data/src/main/resources/component-templates/traces-otel@mappings.yaml:
```diff
@@ -0,0 +1,65 @@
+          unit: nanos
+      dropped_events_count:
+        type: long
+      links:
```
Q: does this handle array of objects properly? Is it already implicitly defined as an object type?

carsonip replied:
> we probably need to set store_array_source so that synthetic source can re-construct in the correct order

quote from @felixbarny

felixbarny replied:
This flag isn't documented, yet but you can read more about it in the PR: https://github.com/elastic/elasticsearch/pull/108417

gregkalapos replied:
Pushed it in https://github.com/elastic/elasticsearch/pull/111091/commits/846db1d13759a5d4edff0ec8647f81969bc54d07 including a test.

gregkalapos committed:
Changing trace_flags to long

Related discussion: https://github.com/elastic/elasticsearch/pull/111091#discussion_r1706698491

axw commented on x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/20_metic_tests.yml:
typo in filename: s/metic/metric/

axw commented on x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/20_traces_tests.yml:
```diff
@@ -0,0 +1,78 @@
+      search:
+        index: traces-generic.otel-default
+        body:
+          fields: ["trace_flags",]
```
```suggestion
          fields: ["trace_flags"]
```

axw commented on x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/20_traces_tests.yml:
```diff
@@ -0,0 +1,78 @@
+        body:
+          fields: ["trace_flags",]
+  - length: { hits.hits: 1 }
+  - match: { hits.hits.0.fields.trace_flags\: [ 4294967295 ] }
```
```suggestion
  - match: { hits.hits.0.fields.trace_flags: [ 4294967295 ] }
```

axw commented on x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/20_traces_tests.yml:
Should there also be a 20_logs_tests.yaml?

gregkalapos replied:
Yes, there will be, but I'd prefer to have it in a follow up. The mappings are mainly driven by our APM POCs and so far we focused on metrics and now started with traces and haven't really done much with logs.

Lots of things are going on in this PR and I feel it'd be beneficial to merge it at some point. This PR moves things into core, so they can be shared across apm and otel - that part is key and independent from the mappings. Also, I feel we won't have the perfect mappings in the initial PR.

Furthermore `otel-data` is disabled by default, so won't really affect users.

So I'd prefer to make a cut here and follow up with logs once we cover that with real use-cases.

axw replied:
:+1: fine by me

axw commented on x-pack/plugin/otel-data/src/main/resources/component-templates/otel@mappings.yaml:
```diff
@@ -0,0 +1,65 @@
+      mode: synthetic
+    properties:
+      "@timestamp":
+        type: date
```
Should we add a TODO comment to update to date_nanos when https://github.com/elastic/elasticsearch/issues/109352 is done?

gregkalapos replied:
Sounds reasonable, added a comment.

axw reviewed (commented):
Looking good, thank you @gregkalapos! I have mostly very minor comments.

gregkalapos committed:
Remove trace_flags

see: https://github.com/elastic/opentelemetry-dev/pull/368#pullrequestreview-2229633970

gregkalapos committed:
Apply suggestions from code review

Co-authored-by: Andrew Wilkins <axwalk@gmail.com>

gregkalapos committed:
Review feedback

gregkalapos commented on x-pack/plugin/otel-data/src/main/resources/component-templates/traces-otel@mappings.yaml:
```diff
@@ -0,0 +1,65 @@
+  managed: true
+template:
+  mappings:
+    date_detection: false
```
We definitely don't need it here 👍 

> And in traces@mappings for that matter, so maybe it doesn't need to be in `otel@mappings` either

That'd affect metrics from what I see. If we remove it from `otel@mappings`, then logs will be still ok, because `logs@mappings` has this as well, but from what I see, for metrics it's not set elsewhere. 

So removed it from here, but kept in `otel@mappings`.

gregkalapos commented on x-pack/plugin/otel-data/src/main/resources/index-templates/metrics-otel@template.yaml:
```diff
@@ -0,0 +1,25 @@
+  description: default OpenTelemetry metrics template installed by x-pack
+  managed: true
+composed_of:
+  - otel@mappings
```
Correct - fixed this.

gregkalapos commented on x-pack/plugin/otel-data/src/main/resources/component-templates/ecs-tsdb@mappings.yaml:
```diff
@@ -0,0 +1,12 @@
+version: ${xpack.oteldata.template.version}
+_meta:
+  description: Default mappings for the OpenTelemetry tsdb index template installed by x-pack
```
Indeed it's orthogonal to OTel and could be reused by other tsdb templates.

So far my strategy was to only move stuff to core when it's shared across multiple templates - just to be on the safe side.

I don't have a strong opinion - both would work. @felixbarny any opinion on this?

axw replied:
Perhaps an even simpler option: remove the component template for now, and define it inline in the `metrics-otel@template` index template. Then if a need to reuse it comes up later, we could extract it into something in core.

felixbarny replied:
The issue with moving this to `metrics-otel@template` is that `metrics-otel@custom` is defined after it. Since the first dynamic template that matches has precedence, user-defined custom templates in `metrics-otel@custom` would be ignored.

Therefore, we need to have a separate component template for these built-in dynamic templates that  go after `metrics-otel@custom`. I don't have a strong opinion on whether `ecs-tsdb@mappings` should be extracted to core. I don't think this component template will get re-used very soon. That's because for integrations, dimensions are defined explicitly. For OTel, all attributes and resource attributes are dimensions and therefore subject to the limitations of dimension fields. This component template takes these limitations into consideration, which is why it's very basic. Due to https://github.com/elastic/elasticsearch/issues/110387, we can't even map `*.ip` fields. So there's not much left of ECS that this component template is mapping.

I'm wondering if we can move the dynamic template to the `metrics-otel@template` index template. We'll need to test if adding dynamic templates to `metrics-otel@custom` will still have precedence.

gregkalapos replied:
> We'll need to test if adding dynamic templates to `metrics-otel@custom` will still have precedence.

I looked into this, but I fail to test this.

With the current status of this PR (so nothing moved yet) I added `metrics-otel@custom` with another dynamic template (just as a test to map strings to integers):
```
"dynamic_templates": [
    {
      "all_strings_to_int": {
        "mapping": {
          "type": "integer"
        },
        "match_mapping_type": "string"
      }
    }
  ]
```

So far from what I see this has no effect - I'd have expected things `"attributes": { "foo": "1",}` to be mapped as a `1` integer.

So I already fail to verify if the current status allows to add dynamic mappings that have precedence over `ecs-tsdb@mappings`.

But I'm a bit confused anyway... What's even the use-case here? In  `ecs-tsdb@mappings` we map strings to keyword, because that's waht TSDB supports (it does not support strings) - so I wonder what could be even overwritten for this in `...@custom`.

In any case - I'm happy to look into this further, but maybe it'd be worth discussing this as a follow up.

felixbarny replied:
> But I'm a bit confused anyway... What's even the use-case here? In  ecs-tsdb@mappings we map strings to keyword, because that's waht TSDB supports (it does not support strings) - so I wonder what could be even overwritten for this in ...@custom.

One example could be that users want to map `*.myipfield` to the `ip` field type. But if we're mapping all strings to keyword fields in a dynamic template that is defined before the @custom component template, users can't do that.

felixbarny replied:
I've added a test, moved the dynamic template to the index template itself, and deleted `ecs-tsdb@mappings`: 16234d7671d1e9785c8f15d177ed632b535a08b5

gregkalapos commented on x-pack/plugin/otel-data/src/main/resources/component-templates/logs-otel@mappings.yaml:
```diff
@@ -0,0 +1,40 @@
+  managed: true
+template:
+  mappings:
+    date_detection: false
```
No - removed 👍

gregkalapos commented on x-pack/plugin/otel-data/src/main/resources/index-templates/README.md:
```diff
@@ -0,0 +1,7 @@
+
+ - `index.default_pipeline`
+ - `index.final_pipeline`
+ - values for `data_stream.*`
```
> Is that an intentional departure from apm-data?

Not really. I think this mostly comes from https://github.com/elastic/elasticsearch/pull/104455 - and honestly I was not aware of this difference. 

The readme file in this form is definitely wrong - we don't set  pipeline related things, and as you point out, `data_stream.*` are set in component templates. So I removed `README.md`

What should we do? Should we move the `data_stream.*` into index templates?

The only thing we'd set is probably `data_stream.type` value. But that's handled in `logs-otel@mappings`, `metrics-otel@mappings`, and in `traces@mappings` (that's in the core). The default value for `data_stream.type` can be already defined in those component templates, so I feel it's a better fit to already define it there, and with that, there isn't really anything to set in the index templates.

So if I understand correctly, this is more like, there is nothing left to set there at this point.

axw replied:
> The only thing we'd set is probably data_stream.type value.

The reason for setting constant values in the index template is to prevent them from being accidentally overridden by `@custom` component templates. The index template takes precedence over component templates, so anything that must not be overridden (e.g. `data_stream.type`) should ideally be defined in the index template.

axw replied:
BTW I wonder if we should fold all of the `<type>-otel@mappings` component templates into their related index templates. There's no reuse going on, so it would be simpler that way. Maybe in a followup?

gregkalapos replied:
> The reason for setting constant values in the index template is to prevent them from being accidentally overridden by @custom component templates. The index template takes precedence over component templates, so anything that must not be overridden (e.g. data_stream.type) should ideally be defined in the index template.

Thanks for pointing this out - that is a really good argument for setting it in the index template. I pushed it in
https://github.com/elastic/elasticsearch/pull/111091/commits/e1adb6ac8acc3481ecc27ce120cb553fa13554a4 and also added some tests to cover this.

> BTW I wonder if we should fold all of the <type>-otel@mappings component templates into their related index templates. There's no reuse going on, so it would be simpler that way. Maybe in a followup?

Maybe the `-otel@custom` component templates would be a challenge as Felix pointed out in https://github.com/elastic/elasticsearch/pull/111091#discussion_r1713397252 - but we can definitely iterate on this, I'd prefer to do so in a follow up.

axw replied:
Sounds good to me, thank you!

gregkalapos commented on x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/OTelPlugin.java:
```diff
@@ -0,0 +1,78 @@
+    // This setting will be ignored if the plugin is disabled.
+    static final Setting<Boolean> OTEL_DATA_REGISTRY_ENABLED = Setting.boolSetting(
+        "xpack.otel_data.registry.enabled",
+        false,
```
The reason for `false` was more that this is super early stage and we wanted to make it opt-in at this point. But adding a comment makes sense - I added one explaining the reasons.

>  staged rollout in serverless, let it soak, then enable by default everywhere

Wasn't aware of this, very useful! Also mentioned this in the comment - I feel we should sync on this and the strategy on when we move to `true` here later.

axw replied:
Thanks for the comment. SGTM, let's chat about it later

gregkalapos commented on x-pack/plugin/otel-data/src/main/resources/resources.yaml:
```diff
@@ -0,0 +1,16 @@
+# "version" holds the version of the templates and ingest pipelines installed
+# by xpack-plugin otel-data. This must be increased whenever an existing template or
+# pipeline is changed, in order for it to be updated on Elasticsearch upgrade.
```
Yeah, no ingest pipelines - was removing a bunch of mentions - this was a leftover.

dakrone reviewed (commented):
I left some comments, and this looks generally fine, but I have to ask, why are these being managed in yaml instead of matching the existing precedent of having them in json? I find yaml much more error-prone for unintentional changes, since it is heavily indentation reliant.

dakrone commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into otel_plugin

axw reviewed (commented):
Basically LGTM, just a few more comments

axw commented:
> why are these being managed in yaml instead of matching the existing precedent of having them in json? I find yaml much more error-prone for unintentional changes, since it is heavily indentation reliant.

@dakrone that's my doing. We have historically had far more issues with business logic (wrong mappings, dynamic templates, etc.) than with syntax (indentation, "Norway problem", etc.). So we're optimising for readability and the ability to comment. Comprehensive (YAML) REST tests will cover basic syntax issues.

gregkalapos committed:
Add store_array_source for span links

gregkalapos committed:
Define constant `data_stream.type` in `template.yaml`s

gregkalapos committed:
Create package-info.java

gregkalapos commented on x-pack/plugin/otel-data/README.md:
```diff
@@ -0,0 +1,33 @@
+## OpenTelemetry Ingest plugin
```
> Maybe we should have both?

As someone who rarely works on this code-base, I immediately looked at readme files, so I think there is benefit to have both.

I added a `package-info.java` as well and kept the readme file.

gregkalapos commented:
> > why are these being managed in yaml instead of matching the existing precedent of having them in json? I find yaml much more error-prone for unintentional changes, since it is heavily indentation reliant.
> 
> @dakrone that's my doing. We have historically had far more issues with business logic (wrong mappings, dynamic templates, etc.) than with syntax (indentation, "Norway problem", etc.). So we're optimising for readability and the ability to comment. Comprehensive (YAML) REST tests will cover basic syntax issues.

Yeah, so I took yaml, because `apm-data` already used yaml and we also discussed with @felixbarny and we said we'd move the json from https://github.com/elastic/elasticsearch/pull/104455 to yaml. I'm personally fine with both, but my 2 cents is that apm and otel should use the same, because it'll be maintained by the same team. So with that I'd have a slight preference to move on with yaml for consistency.

There are already tests here that load these mappings and checks for them, so I'm less worried about accidental wrong indentation.

axw commented on x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/20_logs.tests.yml:
```diff
@@ -0,0 +1,33 @@
+            settings:
+              index:
+                routing_path: [unit, attributes.*, resource.attributes.*]
+                mode: time_series
```
Was this intentional, using TSDS for logs? If so, maybe worth a comment explaining why

strawgate replied:
and if so -- Is it reasonable to default these to be logsdb indices?

felixbarny replied:
Seems to be a copy/paste leftover from the metrics tests.

axw reviewed (approved):
LGTM! Thanks @gregkalapos :pray: Just one minor comment.

@gregkalapos I think you raise a good question in https://github.com/elastic/elasticsearch/pull/111091/files/a6baf4644924ea4929ae2ab58ebac0b78f936e62#r1714175422 - it's also unclear to me if a user should be able to override the `all_strings_to_keywords` dynamic template. I can't think of any reasons.

If it shouldn't be possible to override, then I think it would be a good candidate for going into the index template. Same goes for top-level fields, but we can iterate on that in a followup.

felixbarny committed:
Move ecs-tsdb@mappings to index template

Add test to verify that @custom template can add dynamic templates with a higher precedence

gregkalapos committed:
Update metrics@mappings.json

Remove summary_gauge and summary_counter since they are covered by summary_metrics

gregkalapos committed:
Move clusterService.getClusterSettings().addSettingsUpdateConsumer to registry

gregkalapos commented on x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/OTelPlugin.java:
```diff
@@ -0,0 +1,80 @@
+        if (enabled) {
+            OTelIndexTemplateRegistry registryInstance = registry.get();
+            registryInstance.setEnabled(OTEL_DATA_REGISTRY_ENABLED.get(settings));
+            clusterService.getClusterSettings().addSettingsUpdateConsumer(OTEL_DATA_REGISTRY_ENABLED, registryInstance::setEnabled);
```
Moved this part into the registry, both here and for `APMPlugin` as well. https://github.com/elastic/elasticsearch/pull/111091/commits/9b93b75d07f83f77d8b17b8f556456a7aa17789e

gregkalapos committed:
Fix code-style

axw commented:
@gregkalapos I see you removed summary_gauge & summary_counter, but summary_metrics doesn't have time_series_metric. Should we add that? I think with "counter"?

felixbarny commented:
> @gregkalapos I see you removed summary_gauge & summary_counter, but summary_metrics doesn't have time_series_metric. Should we add that? I think with "counter"?

I'm not sure if it makes much of a difference. I suppose it's relevant for downsampling. I'm not sure if counter or gauge is most appropriate here. The reason we removed summary_gauge & summary_counter is that there's no temporality for summary metrics. It's also not always monotonically increasing as it depends on the start and end timestamp. So you can effectively have either temporality without really having an indication of what the temporality is. At least that's my understanding of it @lahsivjar, please correct me if I'm wrong.

For now, I'm in inclined to just leave out time_series_metric and then potentially add one in a follow-up.

felixbarny commented on x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/20_logs.tests.yml:
```diff
@@ -0,0 +1,33 @@
+  - do:
+      cluster.health:
+        wait_for_events: languid
+  - do:
+      cluster.put_component_template:
+        name: logs-otel@custom
+        body:
+          template:
+            settings:
+              index:
+                routing_path: [unit, attributes.*, resource.attributes.*]
+                mode: time_series
+                time_series:
+                  start_time: 2024-07-01T13:03:08.138Z
```
```suggestion
```

axw commented:
> I'm not sure if it makes much of a difference. I suppose it's relevant for downsampling. I'm not sure if counter or gauge is most appropriate here. The reason we removed summary_gauge & summary_counter is that there's no temporality for summary metrics. It's also not always monotonically increasing as it depends on the start and end timestamp. So you can effectively have either temporality without really having an indication of what the temporality is. At least that's my understanding of it @lahsivjar, please correct me if I'm wrong.
> 
> For now, I'm in inclined to just leave out time_series_metric and then potentially add one in a follow-up.

Sounds very reasonable to me.

gregkalapos committed:
Update x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/20_logs.tests.yml

Co-authored-by: Felix Barnsteiner <felixbarny@users.noreply.github.com>

gregkalapos committed:
Enable logsdb

gregkalapos commented on x-pack/plugin/core/template-resources/src/main/resources/metrics@mappings.json:
```diff
@@ -43,6 +43,68 @@
+          "histogram": {
+            "mapping": {
+              "type": "histogram",
+              "ignore_malformed": true
```
Listed this in https://github.com/elastic/elasticsearch/issues/111877

gregkalapos commented:
I opened https://github.com/elastic/elasticsearch/issues/111877 to collect follow up issues. Feel free to add to it if I missed something.

The new `otel-data` plugin is disabled by default. Since there are multiple things going in this PR - as discussed in other meetings - we should cut here and merge if this is ok to go. With that, other teams can test this more easily. https://github.com/elastic/elasticsearch/issues/111877 will address remaining non-blockers.

carsonip commented on x-pack/plugin/core/template-resources/src/main/resources/traces@settings.json:
```diff
@@ -0,0 +1,21 @@
+    "settings": {
+      "index": {
+        "lifecycle": {
+          "name": "traces"
```
lifecycle is missing

carsonip replied:
The implication of this is that index management page will show a "policy has lifecycle errors" error, as well as the following message when clicking into it.

```
{
  "step_info": {
    "type": "illegal_argument_exception",
    "reason": "policy [traces] does not exist"
  }
}
```

gregkalapos replied:
This is in core, but from what I see it's only used by `otel-data` - we decided to not set lifecycle for `otel-data` yet. And it seems `apm-data` sets lifecycles differently. So from what I see it's safe to remove this here. Basically a copy-paste leftover.

gregkalapos replied:
Removed in https://github.com/elastic/elasticsearch/pull/111091/commits/3bb6737b99f9f3fa89312885607af66aef603904

gregkalapos committed:
Update traces@settings.json

No lifecycle needed for OTel at this point

dakrone reviewed (approved):
LGTM for the Java parts, I can't speak to the content of the mappings (it looks like other Obs folks have already reviewed those parts)

gregkalapos commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into otel_plugin
